app/src/org/commcare/connect/network/base/BaseApi.kt

```diff
@@ -0,0 +1,122 @@
+package org.commcare.connect.network.base
+
+import android.content.Context
+import android.os.Handler
+import okhttp3.ResponseBody
+import org.commcare.activities.CommCareActivity
+import org.commcare.connect.ConnectConstants
+import org.commcare.connect.network.IApiCallback
+import org.commcare.util.LogTypes
+import org.javarosa.core.services.Logger
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.HttpException
+import retrofit2.Response
+import java.io.IOException
+
+class BaseApi {
+
+    companion object {
+        fun callApi(
+            context: Context,
+            call: Call<ResponseBody>,
+            callback: IApiCallback,
+            endPoint: String
+        ) {
+            showProgressDialog(context)
+            call.enqueue(object : Callback<ResponseBody?> {
+                override fun onResponse(
+                    call: Call<ResponseBody?>,
+                    response: Response<ResponseBody?>
+                ) {
+                    dismissProgressDialog(context)
+                    if (response.isSuccessful && response.body() != null) {
+                        // Handle success
+                        try {
+                            response.body()!!.byteStream().use { responseStream ->
+                                callback.processSuccess(response.code(), responseStream)
+                            }
+                        } catch (e: IOException) {
+                            // Handle error when reading the stream
+                            callback.processFailure(response.code(), null, endPoint)
+                        }
+                    } else {
+                        // Handle validation errors
+                        logFailedResponse(response)
+                        val stream = if (response.errorBody() != null) response.errorBody()!!
+                            .byteStream() else null
+                        callback.processFailure(response.code(), stream, endPoint)
+                    }
+                }
+
+                override fun onFailure(call: Call<ResponseBody?>, t: Throwable) {
+                    dismissProgressDialog(context)
+                    // Handle network errors, etc.
+                    logNetworkError(t)
+                    callback.processNetworkFailure()
+                }
+            })
+        }
+
+
+        fun showProgressDialog(context: Context) {
+            if (context is CommCareActivity<*>) {
+                val handler = Handler(context.getMainLooper())
+                handler.post {
+                    try {
+                        (context as CommCareActivity<*>).showProgressDialog(ConnectConstants.NETWORK_ACTIVITY_ID)
+                    } catch (e: Exception) {
+                        //Ignore, ok if showing fails
+                    }
```

Comment:
it does seem weird to have a catch all here for show and not for dimiss below ? Would prefer if we remove this from here as well.